### PR TITLE
osd: add clear_shards_repaired command

### PR DESCRIPTION
**Pull request by diffs**
This command will allow us to clear the `OSD_TOO_MANY_REPAIRS` alert by setting the shard repair count to 0. This will help in cases where the alert was a false positive, or a condition that has since cleared at the disk level. Often, zeroing out the repair count is better than muting the alert or restarting the OSD.

Fixes: https://tracker.ceph.com/issues/54182

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by DanWritesCode** [comment]
FYI this was originally based on commit 1a63a63 from PR #36379 that added this feature to Nautilus, hence the co-authoring.

File: doc/rados/operations/health-checks.rst [hunk 1]
```diff
@@ -983,6 +983,13 @@ the object data, there might exist failing disks that are not registering any
 scrub errors. This repair count is maintained as a way of identifying any such
 failing disks.
 
+In order to allow clearing of the warning, a new command
+``ceph tell osd.# clear_shards_repaired [count]`` has been added.
+By default it will set the repair count to 0.  If the administrator wanted
```
Comment (anthonyeleven) [inline doc/rados/operations/health-checks.rst:988]: s/wanted/wants/

**Review by anthonyeleven — COMMENTED** [review]
docs look okay, one request

File: doc/rados/operations/health-checks.rst (hunk 1 above)
Comment (DanWritesCode) [reply doc/rados/operations/health-checks.rst]: re-worded the paragraph a bit more

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -4350,6 +4362,12 @@ void OSD::final_init()
     asok_hook,
     "debug the scrubber");
   ceph_assert(r == 0);
+  r = admin_socket->register_command(
+    "clear_shards_repaired " \
```
Comment (ronen-fr) [inline src/osd/OSD.cc:4366]: no need for the '\\'

File: src/osd/PGBackend.h [hunk 3]
```diff
@@ -299,6 +299,7 @@ typedef std::shared_ptr<const OSDMap> OSDMapRef;
      virtual bool check_failsafe_full() = 0;
 
      virtual void inc_osd_stat_repaired() = 0;
+     virtual void set_osd_stat_repaired(int64_t) = 0;
```
Comment (ronen-fr) [inline src/osd/PGBackend.h:302]: Why is this needed? unlike inc_osd_stat_repaired(), there is no clearing from the backend

**Review by ronen-fr — CHANGES_REQUESTED** [review]
two minor comments. LGTM otherwise.
Useful for my tests...

**Comment by ronen-fr** [comment]
@diffs : ping?

File: src/osd/OSD.cc (hunk 2 above)
Comment (diffs) [reply src/osd/OSD.cc]: TIL - fixed.

**Comment by diffs** [comment]
Hey @ronen-fr - sorry about the delay, I've pushed an updated commit that addresses your comments.

**Comment by ronen-fr** [comment]
jenkins test api

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -1051,6 +1051,13 @@ void OSDService::inc_osd_stat_repaired()
   return;
 }
 
+void OSDService::set_osd_stat_repaired(int64_t count)
+{
+  std::lock_guard l(stat_lock);
+  osd_stat.num_shards_repaired = count;
+  return;
```
Comment (ronen-fr) [inline src/osd/OSD.cc:1058]: (sorry for missing this in the first round)
Please remove the 'return' line.
(and - if you can - please remove it from the inc_osd_stat_repaired() above, too.)

**Comment by ronen-fr** [comment]
LGTM apart from a minor comment

**Review by ronen-fr — CHANGES_REQUESTED** [review]
an extra 'return' line

File: src/osd/OSD.cc (hunk 4 above)
Comment (diffs) [reply src/osd/OSD.cc]: Done!

**Review by ronen-fr — APPROVED** [review]

**Comment by ronen-fr** [comment]
@diffs - the PR is approved, but fails CI tests, including:
"Signed-off-by — one or more commits in this PR are not signed"

Please sign that commit & repush.

**Review by ronen-fr — CHANGES_REQUESTED** [review]
see latest comment

**Comment by diffs** [comment]
My bad! I had pushed the previous commit too quickly and forgot about the signoff requirement. Now it's fixed. 🙂

**Comment by ljflores** [comment]
jenkins test api

**Comment by diffs** [comment]
@ljflores @ronen-fr just bumping this PR up due to the stale message

**Comment by frittentheke** [comment]
@ljflores @ronen-fr @anthonyeleven why is this PR staling out? Any chance this could go to the list for the next release / QA runs? Last comment in this direction was https://github.com/ceph/ceph/pull/54954#issuecomment-1989133704 by @ljflores indicating that it should have gone towards testing?

**Comment by olifre** [comment]
Since this is staling out again — I believe all changes requested by @ronen-fr have been applied, correct?

**Comment by diffs** [comment]
@olifre Ronen approved the changes on 3/5. Then requested that the commits be signed and force-pushed on 3/6. This was done on 3/6. The 2 commits were not squashed into 1 commit, though. Would it help if I did that?

**Comment by ronen-fr** [comment]
LGTM, apart from a question:
The PR uses stat_lock. Is it enough, or is full_status_lock required?
@diffs do you know the answer?
if not: @rzarzynski ?

**Comment by diffs** [comment]
@ronen-fr FWIW it uses the same lock type that `inc_osd_stat_repaired()` uses, which is a much older function. `stat_lock` sounds right to me, since it's just updating an OSD stat, but beyond that I'm not sure.

**Review by ronen-fr — APPROVED** [review]

**Comment by amathuria** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues68375
